Add previous-image flag

importhuman opened the pull request:
Signed-off-by: Ujjwal Goyal <importujjwal@gmail.com>

## Summary
<!-- Provide a high-level summary of the change. -->
Add `previous-image` flag for #897 with an `Images` field. I have started with the changes required to the `analyze` phase, but am having trouble writing the test for it (passing when it should fail). 

Additional assistance required:
- cannot run `make format`, `make tidy` or `make prepare-for-pr`, getting the error `"./create_builder.go" contains more than 3 groups of imports`.
- observing some untracked files with the name format "internal/registry/registry177994269/" while making changes, what to do with them? 
- unclear about when to run `make prepare-for-pr`. Does "before creating a Pull Request" mean before adding and committing?

## Output
<!-- If applicable, please provide examples of the output changes. -->

#### Before

#### After

## Documentation
<!-- If this change should be documented, please create an issue or PR on https://github.com/buildpacks/docs and link below. -->
<!-- NOTE: This can be added (by editing the issue) after the PR is opened. -->

- Should this change be documented?
    - [ ] Yes, see #897 
   
## Related
<!-- If this PR addresses an issue, please provide issue number below. -->

Concerns #897 (Doesn't resolve yet)

aemengo commented:
Thanks so much for helping out!

> getting the error "./create_builder.go" contains more than 3 groups of imports.

Wild. I never see that error, but the linter is correct. The imports should be: **stdlib**, **external dependencies**, **internal dependencies**. You didn't introduce the error as part of this PR so I don't think it's worth your time, _in this PR_. See: https://github.com/buildpacks/pack/pull/1187#discussion_r634624030

> observing some untracked files with the name format "internal/registry/registry177994269/" while making changes, what to do with them?

Please just ignore them. We definitely need to cleanup, our cleanup logic there. 😅

> unclear about when to run make prepare-for-pr. Does "before creating a Pull Request" mean before adding and committing?

After adding and committing, but before _git push_.

aemengo commented on build.go:
```diff
@@ -159,6 +159,11 @@ type BuildOptions struct {
 	// User's group id used to build the image
 	GroupID int
+
+	// Slice of two images
+	// 1. A previous image to set to a particular tag reference, digest reference, or image ID;
+	// 2. The image to refer to
+	Images []string
```
I'm a little concerned about the direction that this PR is headed.. The previous image flag is intended to have space for one `string`, not two. If you see https://github.com/buildpacks/pack/blob/5ea76488f9c16bca671dd417da1f311f423a7d8b/build.go#L72, you'll see where that `<image-name>` was _already_ being collected as part of the pack build command.

Happy to clarify, if I wasn't clear!

aemengo replied:
Forgive me, I'm having trouble understanding.

If you're asking how do I know `PreviousImage` was passed in by a user? You could assert that the value is not equal to an empty string (`""`). 

If you wanna talk about these things in a more realtime way, we're all pretty active at: https://slack.buildpacks.io/

dfreilich reviewed (commented):
In terms of code, @importhuman , this looks superb. Fits with our coding styles, and is very clean and pristine. Good work! 

In terms of UX, I'm a bit concerned that people may not understand exactly what this option is (I would have difficulty explaining when to use this or not to use it). @ekcasey  or @natalieparellano , can we have some further explanation of how/why people would want to use this, which we can add to the flag description and make it a bit clearer?

importhuman commented:
@dfreilich Thanks! Trying to figure out the edge case now. Also, not working on the analyzer scenario in this PR, as suggested by @natalieparellano on slack.

I don't really understand how to use myself either, I've been going along what was given in the description and further discussions. 😅

importhuman commented on acceptance/acceptance_test.go:
```diff
@@ -1764,6 +1764,45 @@ func testAcceptance(
 						})
 					})
 
```
Couldn't figure out how to write the acceptance test. This is based off the cache-image test, but failing.

aemengo replied:
Let's not bother with an acceptance test for this.

They're slower, require more resources, and are more likely to fail. We should reserve them for more critical code paths with more moving parts. Although this PR touches a few moving parts, we have plenty other acceptance tests that the reduce need for this one.

importhuman replied:
@aemengo I did observe that it took a bit of time. I added the acceptance test because it was suggested to do so in the [issue](https://github.com/buildpacks/pack/issues/897#issuecomment-853059556). If @jromero says it's alright without this test then I agree that we can skip it.

importhuman commented on internal/build/lifecycle_execution_test.go:
```diff
@@ -800,20 +800,95 @@ func testLifecycleExecution(t *testing.T, when spec.G, it spec.S) {
 
-				lastCallIndex := len(fakePhaseFactory.NewCalledWithProvider) - 1
-				h.AssertNotEq(t, lastCallIndex, -1)
+			// when("image is invalid", func() {
```
Tried to add a unit test to check for invalid images, but couldn't figure out how to pass an invalid image name in the test as it gets validated. Tried passing tags like "%%%", tags with other invalid characters, tags with multiple slashes, etc.

aemengo replied:
Once again, I'd rather not bother with a test for this 😅 

A test to check that we're bubbling up the error would be of _low value_, imo. Such a test would be more valuable if it were _our logic_ deciding what a valid/invalid image tag is.

In this case, I'd be comfortable having faith in our third-party library having its own tests. And then saving ourselves the keystrokes for every error check.

importhuman replied:
Cool, will remove.

importhuman commented on internal/commands/build_test.go:
```diff
@@ -709,6 +709,40 @@ builder = "my-builder"
+						h.AssertNil(t, command.Execute())
+					})
+				})
+
```
Couldn't figure out this test case, get the following error message:
`=== RUN   TestBuildCommand/Commands/#BuildCommand/previous-image_flag_is_provided/--publish_is_true/image_and_previous-image_are_not_in_same_registry/fails
    build.go:130: Unexpected call to *testmocks.MockPackClient.Build([context.Background { index.docker.io/some/image:latest my-builder     map[] map[]  true false false [] [] <nil> { []}  always . {{ []} {[] [] [] [] } map[]}   -1 example.io/some/previous-image}]) at /Users/ujjwal/coding/pack/internal/commands/build.go:130 because: there are no expected calls of the method "Build" for that receiver`

aemengo replied:
Mock assertions are always painful to read. In this case, it's complaining that your mock received a call, but your _test setup_ did not have the relevant `mockClient.EXPECT()` line.

Stepping back for a second. This unit test doesn't feel appropriate here. The logic for actually creating the error is way down in `internal/build/lifecycle_execution.go`. It feels like there's a decision to make about moving the test down or bringing the logic up...

I'd much prefer that you bring this logic (_and tests_) up. In fact, I'd prefer them to be at: **internal/commands/build.go**. That way we don't unnecessarily pull images, perform downloads, etc and then the action will fail anyway.

I'm glad that you tackled the edge case for us though 🙂

importhuman replied:
Again, I added the test here because it was mentioned in the [issue](https://github.com/buildpacks/pack/issues/897#issuecomment-853059556). 😅 

I believe I can add the logic to `internal/commands/build.go`, but I don't get how to move the tests. Aren't the tests for this file in `internal/commands/build_test.go`, which is the file we're talking about here? Or would the logic of checking for invalid previous-image passed remove the need for a test? 😅 

cc: @jromero 

But for the sake of better understanding, I'd like to understand the error message.
> Mock assertions are always painful to read. In this case, it's complaining that your mock received a call, but your test setup did not have the relevant mockClient.EXPECT() line.

When I have the `mockClient` line, I got the error "expected error but got nil". I tried it with various registry names, but got the same result. For example, with previous image name `example.io/some/previous-image` and image name `index.docker.io/some/image:latest`. Apologies if this information is insufficient.

Also, thank you so much for helping me with this PR, really appreciate it 😄

jromero replied:
Just sent you a DM on slack if you want to do a bit of pair programing to get past this little hurdle.

importhuman reviewed (commented):
Having some trouble with writing certain tests 😅

importhuman commented:
Some tests are failing in the `TestLifecycleExecution` suite, but none of the failing tests are ones I've added. I updated my local branch with the latest changes in the main branch, could this be the reason?
cc: @jromero

importhuman commented:
@jromero How can I fix the tests failing at the "no test files" stage?  Other than this, the PR should be good to go. I've put back the image validation portion inside the condition as it was before, otherwise the tests were failing since not all of them were provided with images.
